### PR TITLE
Warn if a new flow is being created that is similar to an existing

### PR DESCRIPTION
**Pull request by kamransaleem**
#1529

### CODE REVIEW

File: waltz-ng/client/physical-flows/physical-flow-registration.html [hunk 1]
```diff
@@ -81,6 +81,48 @@
...
+                <h4>
+                    Existing flows from
+                    <waltz-entity-icon-label entity-ref="ctrl.sourceEntity"></waltz-entity-icon-label>
+                    to
+                    <waltz-entity-icon-label entity-ref="ctrl.targetEntity"></waltz-entity-icon-label>
+                </h4>
+                <waltz-grid column-defs="ctrl.similarFlowDefs"
+                            row-data="ctrl.similarFlows"
+                            class="small">
+                </waltz-grid>
```
Comment (davidwatkins73) [inline waltz-ng/client/physical-flows/physical-flow-registration.html]: Using a grid is a bit excessive - they are fairly heavyweight components.....

File: waltz-ng/client/physical-flows/physical-flow-registration.js [hunk 2]
```diff
@@ -72,6 +75,17 @@ function validate(specification, flowAttributes, targetEntity) {
 }
 
 
+function getSimilarFlows(specification, flowAttributes, targetEntity, existingFlows = []) {
```
Comment (davidwatkins73) [inline waltz-ng/client/physical-flows/physical-flow-registration.js]: minor:  we tend to name things `findXXX` if we expect to receive a collection.

Comment (davidwatkins73) [inline waltz-ng/client/physical-flows/physical-flow-registration.js]: Whats the purpose of passing in `flowAttributes` and `targetEntity` - they don't seem to be used other than in the `if`

File: waltz-ng/client/physical-flows/physical-flow-registration.js [hunk 3]
```diff
@@ -93,7 +119,12 @@ function controller(
     const viewState = kindToViewState(sourceEntityRef.kind);
     vm.cancelLink = $state.href(viewState, { id: sourceEntityRef.id });
 
-    const doValidation = () => validate(vm.specification, vm.flowAttributes, vm.targetEntity);
+    const doValidation = () => {
+        vm.similarFlows = [];
```
Comment (davidwatkins73) [inline waltz-ng/client/physical-flows/physical-flow-registration.js]: why initialise here if it's about to be blatted below ?

File: waltz-ng/client/physical-flows/physical-flow-registration.js [hunk 4]
```diff
@@ -93,7 +119,12 @@ function controller(
...
+        vm.visibility.similarFlows = false;
```
Comment (davidwatkins73) [inline waltz-ng/client/physical-flows/physical-flow-registration.js]: not sure this adds a lot, why not simply test the length of `similarFlows` - having independent derived state can lead to subtle bugs.

File: waltz-ng/client/physical-flows/physical-flow-registration.js [hunk 5]
```diff
@@ -93,7 +119,12 @@ function controller(
...
+        vm.similarFlows = getSimilarFlows(vm.specification, vm.flowAttributes, vm.targetEntity, vm.existingFlowsToTarget);
+        return validate(vm.specification, vm.flowAttributes, vm.targetEntity);
```
Comment (davidwatkins73) [inline waltz-ng/client/physical-flows/physical-flow-registration.js]: this is returning an object which is then ignored later.

File: waltz-ng/client/physical-flows/physical-flow-registration.js [hunk 6]
```diff
@@ -114,6 +145,11 @@ function controller(
 
     vm.targetChanged = (target) => {
         vm.targetEntity = target;
+        physicalFlowStore.findByConsumerEntityReference(target)
+            .then(flows => {
+                vm.existingFlowsToTarget = flows;
+                doValidation();
```
Comment (davidwatkins73) [inline waltz-ng/client/physical-flows/physical-flow-registration.js]: this is the bit that ignores the return of `doValidation`

File: waltz-ng/client/physical-flows/physical-flow-registration.js (hunk 2 above)
Comment (kamransaleem) [inline waltz-ng/client/physical-flows/physical-flow-registration.js]: Want the check for similar flows to happen at the end once the user has specified all flow attributes.  I figured any earlier would interrupt the flow.

will rename to findSimilarFlows

File: waltz-ng/client/physical-flows/physical-flow-registration.js (hunk 6 above)
Comment (kamransaleem) [inline waltz-ng/client/physical-flows/physical-flow-registration.js]: agreed, will remove the return

File: waltz-ng/client/physical-flows/physical-flow-registration.js (hunk 5 above)
Comment (kamransaleem) [inline waltz-ng/client/physical-flows/physical-flow-registration.js]: will remove return

File: waltz-ng/client/physical-flows/physical-flow-registration.js (hunk 4 above)
Comment (kamransaleem) [inline waltz-ng/client/physical-flows/physical-flow-registration.js]: this is so we can have the 'show' button controlling whether or not those flows are visible.

File: waltz-ng/client/physical-flows/physical-flow-registration.html (hunk 1 above)
Comment (kamransaleem) [inline waltz-ng/client/physical-flows/physical-flow-registration.html]: Primarily to keep the styling between this and other areas where physical flows are in a table.  Can change to a <table> instead?

File: waltz-ng/client/physical-flows/physical-flow-registration.js (hunk 2 above)
Comment (davidwatkins73) [inline waltz-ng/client/physical-flows/physical-flow-registration.js]: Perhaps make it more explict:
```
if (! a || ! b)  { return [] };
if (c) { doWork() }
else { return [] }
```

File: waltz-ng/client/physical-flows/physical-flow-registration.html [hunk 7]
```diff
@@ -81,6 +81,48 @@
 
         </div>
 
+
+        <div ng-if="ctrl.visibility.editor == '' && ctrl.similarFlows.length > 0">
+            <div class="row" ng-hide="ctrl.visibility.similarFlows">
```
Comment (davidwatkins73) [inline waltz-ng/client/physical-flows/physical-flow-registration.html]: re earlier comment about `visibility.similarFlows` vs `ctrl.similarFlows` we have both flavours of the test here....

Don't think the extra derived flag is adding a lot.

**Pull request merged**